### ISSUE TITLE
grunt test / karma:unit does not use the latest code

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#2587]
`grunt test` invokes `karma:unit`, which uses code in `dist/inbox.js`. However, the code is only compiled into that file when browserify is run, which is only on the default or dev grunt commands.

We should either:
- Change karma to the actual code source files in `static/js/` instead of `static/dist`. If we're worried about browserify creating issues we could have `grunt ci` run against `static/dist` (ie on travis).
- Change `grunt test` to run browserify etc against the code before running tests

I'd prefer the former, as we'd also get useful stack traces in errors, instead of `inbox.js:232452`.

**Comment by alxndrsn** [comment]
Option · (the first one) sounds better to me too, mainly for the reason you state - tracking down the line that's causing a test failure is currently not easy.

**Comment by garethbowen** [comment]
I agree option one sounds fantastic, but I think it would be very hard to implement because browsers don't understand `require` statements. That's why we run `browserify` so that we end up with something the browser will understand.

In my dev environment I run `grunt dev` constantly, so any changes are compiled into dist right away. This works well for me, but it is annoying if I forget it and wonder why my tests are still failing. It would also be annoying if browserify was run in `grunt dev` and then run when I `grunt test` as it takes a while...

The gold plated solution would probably be to try and detect when dist is out of date. But it may be enough to have a flag or something to instruct grunt to build then test. Or maybe have separate grunt tasks - one for test and one for build then test?

**Comment by SCdF** [comment]
Yay, that worked, both 5.x and 6.x pass in Travis. Now to add 0.12 back…

**Comment by nice-snek** [comment]
hi frand @${assignee}, U is coot

Please close or schedule before the end of this sprint. See [triaging old issues](https://github.com/medic/medic-docs/blob/master/md/dev/workflow.md#triaging-old-issues).

Nice Snek

**Comment by nice-snek** [comment]
hi frand @alxndrsn, please [triage](https://github.com/medic/medic-docs/blob/master/development/workflow.md#triaging-old-issues) before the end of this sprint.

**Comment by SCdF** [comment]
Closing this. If it annoys people enough we can take a deeper look at it another time.

**Comment by alxndrsn** [comment]
This annoys me every time I run tests locally.  I'll look into it sometime.

**Comment by alxndrsn** [comment]
Running `grunt dev` locally doesn't actually seem to solve this issue for me.

Instead I end up running e.g. `grunt mmjs karma`, but only after I've wasted some time re-discovering that these targets don't run against the latest code.

I still want to find a fix for this.

**Comment by garethbowen** [comment]
Here's a possible solution for mocking angular code so we can run these tests in mocha in node without browserifying everything together: https://gist.github.com/rikukissa/dcb422eb3b464cc184ae

**Comment by dianabarsan** [comment]
@garethbowen I think it's worth a try.
